pkg/cmd/server/crypto/crypto.go

```diff
@@ -0,0 +1,475 @@
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"text/template"
+	"time"
+
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/golang/glog"
+)
+
+func filenamesFromDir(dir string) (string, string, string) {
+	if absdir, err := filepath.Abs(dir); err == nil {
+		dir = absdir
+	}
+	return filepath.Join(dir, "root.crt"), filepath.Join(dir, "cert.crt"), filepath.Join(dir, "key.key")
+}
+
+type TLSCertificateConfig struct {
+	CAFile   string
+	CertFile string
+	KeyFile  string
+
+	Roots []*x509.Certificate
+	Certs []*x509.Certificate
+	Key   crypto.PrivateKey
+}
+
+func (c *TLSCertificateConfig) writeDir(dir string) error {
+	c.CAFile, c.CertFile, c.KeyFile = filenamesFromDir(dir)
+
+	// mkdir
+	if err := os.MkdirAll(dir, os.FileMode(0755)); err != nil {
+		return err
+	}
+
+	// write certs and keys
+	if err := writeCertificates(c.CAFile, c.Roots...); err != nil {
+		return err
+	}
+	if err := writeCertificates(c.CertFile, c.Certs...); err != nil {
+		return err
+	}
+	if err := writeKeyFile(c.KeyFile, c.Key); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newTLSCertificateConfig(dir string) (*TLSCertificateConfig, error) {
+	caFile, certFile, keyFile := filenamesFromDir(dir)
+	config := &TLSCertificateConfig{
+		CAFile:   caFile,
+		CertFile: certFile,
+		KeyFile:  keyFile,
+	}
+
+	if caFile != "" {
+		caPEMBlock, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		config.Roots, err = certsFromPEM(caPEMBlock)
+		if err != nil {
+			return nil, fmt.Errorf("Error reading %s: %s", caFile, err)
+		}
+	}
+
+	if certFile != "" {
+		certPEMBlock, err := ioutil.ReadFile(certFile)
+		if err != nil {
+			return nil, err
+		}
+		config.Certs, err = certsFromPEM(certPEMBlock)
+		if err != nil {
+			return nil, fmt.Errorf("Error reading %s: %s", caFile, err)
+		}
+
+		if keyFile != "" {
+			keyPEMBlock, err := ioutil.ReadFile(keyFile)
+			if err != nil {
+				return nil, err
+			}
+			keyPairCert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+			if err != nil {
+				return nil, err
+			}
+			config.Key = keyPairCert.PrivateKey
+		}
+	}
+
+	return config, nil
+}
+
+func writeClientConfigToDir(client *kclient.Config, dir string) error {
+	// mkdir
+	if err := os.MkdirAll(dir, os.FileMode(0755)); err != nil {
+		return err
+	}
+
+	// write files
+	client.CAFile, client.CertFile, client.KeyFile = filenamesFromDir(dir)
+	if err := ioutil.WriteFile(client.CAFile, client.CAData, os.FileMode(0644)); err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(client.CertFile, client.CertData, os.FileMode(0644)); err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(client.KeyFile, client.KeyData, os.FileMode(0600)); err != nil {
+		return err
+	}
+
+	t := template.Must(template.New("kubeconfig").Parse(`
```

Comment:
Minor: In theory the serialization is stable (backwards compatible), but it might be worth building up the objects and serializing the object out.